### DISCUSSION TITLE
Milvus Cluster with Docker Compose on single machine okay for production?

### DISCUSSION THREAD

**Discussion by schuberttobias** [milvus-io/milvus#20916] [Q&A and General discussion] [answered]
Hi!

Since June 2022 Milvus documentation says 

> Docker Compose cannot deploy Milvus distributed clusters across machines, and can only be used in test environments. It is not recommended that you deploy Milvus distributed clusters in this way in production environments. 

## Questions
1. What about Milvus Cluster with Docker Compose on **a single machine for production**: Is that supported, recommended and stable? 
2. Or do you also recommend K8s for Milvus Cluster for production usage, even if only a single machine is used? 
3. What are the drawbacks of using Milvus Cluster with Docker Compose - other than not being able to use more than 1 machine?

## Background
- Even on a single machine, Milvus Cluster v2.0.2 has brought us a [significant performance increase over Milvus Standalone](https://github.com/milvus-io/milvus/discussions/16867#discussioncomment-4208501).
  - Has this changed with Milvus 2.2? Is Milvus Standalone now faster than Milvus Cluster on a single machine? @yhmo advises to use Standalone instead of Cluster on a single machine (see https://github.com/milvus-io/milvus/discussions/18224#discussioncomment-3119936).

**Comment by yhmo** [comment, marked as answer]
1. What about Milvus Cluster with Docker Compose on a single machine for production: Is that supported, recommended and stable?
Supported but not recommended. If we deploy a cluster in a single machine, all the components share the CPU and memory, it is not easy to control the memory usage. If both indexnode and querynode are working, index process will significantly affect the performance of search.


2. Or do you also recommend K8s for Milvus Cluster for production usage, even if only a single machine is used?
We suppose if user uses k8s to deploy, there should be not only one machine in the cluster.


3. What are the drawbacks of using Milvus Cluster with Docker Compose - other than not being able to use more than 1 machine?
The reason is answered in question 1.

**Reply by schuberttobias** [reply]
Hi @yhmo,

thanks for your help!

> If both indexnode and querynode are working, index process will significantly affect the performance of search.

This is a very interesting point. Actually, we are experiencing exactly this problem currently (Milvus v2.0.2 in Docker Compose with 3 proxies, 3 query nodes and 1 index node). During indexing, search performance degrades from < 1 second per query to ~2-4 seconds per query.

We will look into migrating from Docker Compose to K8s.
